OutOfSchool/OutOfSchool.WebApi/Controllers/V1/AdminController.cs

```diff
@@ -1,44 +1,274 @@
+using OutOfSchool.WebApi.Models.Application;
+using OutOfSchool.WebApi.Models.Providers;
 
 namespace OutOfSchool.WebApi.Controllers.V1;
 
 [ApiController]
 [ApiVersion("1.0")]
 [Route("api/v{version:apiVersion}/[controller]/[action]")]
-public class AdminController : ControllerBase
+public class AdminController : Controller
 {
-    private readonly IUserService userService;
-    private readonly IMapper mapper;
+    private readonly IStringLocalizer<SharedResource> localizer;
 
-    public AdminController(IUserService userService, IMapper mapper)
+    private readonly ILogger<AdminController> logger;
+
+    private readonly ICurrentUserService currentUserService;
+    private readonly ISensitiveMinistryAdminService ministryAdminService;
+    private readonly ISensitiveDirectionService directionService;
+    private readonly ISensitiveProviderService providerService;
+    private readonly ISensitiveApplicationService applicationService;
+
+
+    public AdminController(
+       ILogger<AdminController> logger,
+       ICurrentUserService currentUserService,
+       ISensitiveMinistryAdminService ministryAdminService,
+       ISensitiveApplicationService applicationService,
+       ISensitiveDirectionService directionService,
+       ISensitiveProviderService providerService,
+       IStringLocalizer<SharedResource> localizer)
+    {
+        this.localizer = localizer;
+        this.logger = logger;
+        this.applicationService = applicationService;
+        this.directionService = directionService;
+        this.providerService = providerService;
+        this.currentUserService = currentUserService ?? throw new ArgumentNullException(nameof(currentUserService));
+        this.ministryAdminService = ministryAdminService ?? throw new ArgumentNullException(nameof(ministryAdminService));
+    }
+
+    /// <summary>
+     /// Get MinistryAdmins that match filter's parameters.
+     /// </summary>
+     /// <param name="filter">Entity that represents searching parameters.</param>
+     /// <returns><see cref="SearchResult{MinistryAdminDto}"/>, or no content.</returns>
+    [HasPermission(Permissions.MinistryAdminRead)]
+    [HttpGet]
+    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(SearchResult<MinistryAdminDto>))]
+    [ProducesResponseType(StatusCodes.Status204NoContent)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(StatusCodes.Status403Forbidden)]
+    [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+    public async Task<IActionResult> GetByFilterMinistryAdmin([FromQuery] MinistryAdminFilter filter)
+     {
+         if (!User.IsInRole(nameof(Role.TechAdmin).ToLower()))
+         {
+             logger.LogError("You have no rights because you are not an admin");
+             return StatusCode(403, "Forbidden to update another user if you don't have TechAdmin role.");
+         }
```

Comment:
> (Code duplication) Please consider better way to enforce user rights checking. Pay attention, that your copy-paste approach is really bad. I.e., check the next:
> 
> ```
>     public async Task<ActionResult> DeleteDirectionById(long id)
>     {
>         if (!User.IsInRole(nameof(Role.TechAdmin).ToLower()))
>         {
>             ...
>             return StatusCode(403, "Forbidden to update another user if you don't have TechAdmin role.");
>         }
> ```
> 
> Method is `DeleteDirectionById` - log text says `Forbidden to update another user`. I believe you could do it better.

I`ve created a private method ```IsTechAdmin``` to decrease duplication)
I hope you like this approach)

Reply:
Though I'll treat this change from you as approved, please check other options: [action filters](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/filters?view=aspnetcore-8.0#action-filters) and [custom authorization policy](https://learn.microsoft.com/en-us/aspnet/core/security/authorization/iard?view=aspnetcore-8.0).

Former is better because it's impossible not execute the `action filter` (in your code you could forget to call `IsTechAdmin()` in some action), and also it is already used in the project (i.e. see the `AreaAdminController`). At same time it is not reusable outside of the controller.

Latter is better, b/c it is reusable outside of one controller, but is worse, b/c you need to write the attribute it in every action (and could forget to do it).